src/main/scala/diplomacy/Nodes.scala

```diff
@@ -259,27 +733,157 @@ trait OutwardNode[DO, UO, BO <: Data] extends BaseNode
+    * preventing subsequent bindings from being created via [[oPush]].
+    *
+    * The bindings are each a tuple of:
+    * - numeric index of this binding in the other end of [[InwardNode]].
+    * - [[InwardNode]] on the other end of this binding
+    * - [[NodeBinding]] describing the type of binding
+    * - a [[Parameters]] instance
+    * - [[SourceInfo]] for source-level error reporting
+    */
+  protected[diplomacy] lazy val oBindings: Seq[(Int, InwardNode[DO, UO, BO], NodeBinding, Parameters, SourceInfo)] = { oRealized = true; accPO.result() }
 
+  /** resolved STAR binding of outward nodes: how many connections the star represents. */
   protected[diplomacy] val oStar: Int
+
+  /** a map to convert Node binding index to port range.
+    *
+    * index: the index of connected [[InwardNode]]
+    * element: port range of connected [[InwardNode]]
+    * */
   protected[diplomacy] val oPortMapping: Seq[(Int, Int)]
+
+  /** "Forward" an output connection through this node so that the node can be removed from the graph.
+    *
+    *  @return None if no forwarding is needed.
+    */
   protected[diplomacy] def oForward(x: Int): Option[(Int, OutwardNode[DO, UO, BO])] = None
-  protected[diplomacy] val uoParams: Seq[UO] // from connected nodes
-  protected[diplomacy] val doParams: Seq[DO] // from this node
+  /** Upstream outward parameters.
+    *
+    * Generated from the other nodes that this is connected to.
+    * For specific diagram, see Parameter propagation flow on the top.
+    */
+  protected[diplomacy] val uoParams: Seq[UO]
+  /** Downstream outward parameters.
+    *
+    * Generated from this node.
+    * For specific diagram, see Parameter propagation flow on the top.
+    */
+  protected[diplomacy] val doParams: Seq[DO]
 }
 
-abstract class CycleException(kind: String, loop: Seq[String]) extends Exception(s"Diplomatic ${kind} cycle detected involving ${loop}")
+abstract class CycleException(kind: String, loop: Seq[String]) extends Exception(s"Diplomatic $kind cycle detected involving $loop")
 case class StarCycleException(loop: Seq[String] = Nil) extends CycleException("star", loop)
 case class DownwardCycleException(loop: Seq[String] = Nil) extends CycleException("downward", loop)
 case class UpwardCycleException(loop: Seq[String] = Nil) extends CycleException("upward", loop)
 
+/** [[Edges]] is a collection of parameters describing the functionality and connection for an interface,
+  * which is often derived from the interconnection protocol and can inform the parameterization
+  * of the hardware Bundles that actually implement the protocol.
+  */
 case class Edges[EI, EO](in: Seq[EI], out: Seq[EO])
+
+/** The sealed Node in the package, all node are derived from it.
+  *
+  * @param inner   sink interface implementation.
+  * @param outer   source interface implementation.
+  * @param valName val name of this node.
+  * @tparam DI Downwards Parameters received on the inner side of the node.
```

Comment:
"Downwards-flowing"?

Definitely don't capitalize "Parameters" cause that means something special...